File: stickerapp/ViewController.m
//  I'ma Unicorn
//
//  Created by Franky Aguilar on 7/27/12.
//  Copyright (c) 2012 99centbrains Inc. All rights reserved.
//  @99centbrains - http://99centbrains.com
//  ALL ARTWORK AND DESIGN OWNED BY 99centbrains, not for reproduction or redistribution
//
#import "ViewController.h"
#import "PlayViewController.h"
#import "AppDelegate.h"
#import "CWInAppHelper.h"
#import <AVFoundation/AVFoundation.h>
#import <AudioToolbox/AudioToolbox.h>
#import "SVProgressHUD.h"
#import "SVModalWebViewController.h"

@interface ViewController (){
    

}


@property (nonatomic, strong)UIPopoverController *popController;


@end


@implementation ViewController

@synthesize ibo_getphoto = _ibo_getphoto;

- (void)viewDidLoad {
   
    [super viewDidLoad];
   
}

- (void)viewDidAppear:(BOOL)animated{
    
    [super viewDidAppear:animated];

    NSLog(@"View Did Appear");
}



//PHOTOS
- (IBAction)iba_photoTake:(id)sender{

        
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    picker.allowsEditing = YES;
    
    if ([UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera]){
        
        
		picker.sourceType = UIImagePickerControllerSourceTypeCamera;
		[self presentViewController:picker animated:YES completion:nil];
        
        
	} else {
        
		[self iba_photoChoose:sender];
        
	}
    
}



- (IBAction)iba_photoChoose:(id)sender{

    
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    picker.allowsEditing = YES;
    
    
    picker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
   
    
    
        
        [self presentViewController:picker animated:YES completion:nil];
        
    
    
}


- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    
	if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad){
        
        
        [_popController dismissPopoverAnimated:YES];
        
    } else {
        
                
    }
    [picker dismissViewControllerAnimated:YES completion:nil];

}

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info{
    
    UIStoryboard *sb = [UIStoryboard storyboardWithName:@"MainStoryboard" bundle:nil];
    
    PlayViewController *playViewController = (PlayViewController *)[sb instantiateViewControllerWithIdentifier:@"seg_PlayViewController"];

    
    [picker dismissViewControllerAnimated:YES completion:^{
        
        playViewController.userImage = [self editedImageFromMediaWithInfo:info];
        [self.navigationController pushViewController:playViewController animated:YES];
        //
    }];

}

- (IBAction)iba_shareApp:(UIButton *)sender{

    
    NSString *shareText = kShareDescription;
    NSURL *shareURL = [NSURL URLWithString:kShareURL];
    
    NSArray *activityItems = @[shareText, shareURL];
    NSArray *excludeActivities = @[UIActivityTypeAddToReadingList,
                                   UIActivityTypeAssignToContact,
                                   UIActivityTypePrint];
    
    UIActivityViewController *activityController = [[UIActivityViewController alloc]
                                                    initWithActivityItems:activityItems
                                                    applicationActivities:nil];
    activityController.excludedActivityTypes = excludeActivities;
    
    //CHECK FOR ACTIVITY TYPE AND DISPLAY CORRECT UIDOCUMENTATIONCONTROLLER
    activityController.completionHandler = ^(NSString *activityType, BOOL completed){
        
    };
    
    if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
        
        _popController = [[UIPopoverController alloc] initWithContentViewController:activityController];
        _popController.popoverContentSize = CGSizeMake(500, 425);
        [_popController presentPopoverFromRect:[sender convertRect:sender.bounds toView:self.view] inView:self.view permittedArrowDirections:UIPopoverArrowDirectionUp animated:YES];
        [SVProgressHUD dismiss];
    
    } else {
        
        [SVProgressHUD dismiss];
        [self presentViewController:activityController animated:YES completion:nil];
    
    }

}

- (IBAction)iba_Facebook:(id)sender{
    
    SVModalWebViewController *modalWeb = [[SVModalWebViewController alloc] initWithURL:[NSURL URLWithString:@"http://facebook.com/99centbrains"]];
    [self presentViewController:modalWeb animated:YES completion:^{
        //
    }];

}

- (IBAction)iba_Twitter:(id)sender{
    
    SVModalWebViewController *modalWeb = [[SVModalWebViewController alloc] initWithURL:[NSURL URLWithString:@"http://twitter.com/catgnaw"]];
    [self presentViewController:modalWeb animated:YES completion:^{
        //
    }];

}

- (IBAction)iba_Insta:(id)sender{
    
    SVModalWebViewController *modalWeb = [[SVModalWebViewController alloc] initWithURL:[NSURL URLWithString:@"http://instagram.com/catgnaw"]];
    [self presentViewController:modalWeb animated:YES completion:^{
        //
    }];

}

- (IBAction)iba_Web:(id)sender{
    
    SVModalWebViewController *modalWeb = [[SVModalWebViewController alloc] initWithURL:[NSURL URLWithString:@"http://catgnaw.com"]];
    [self presentViewController:modalWeb animated:YES completion:^{
        //
    }];

}

//VIEW UNLOAD
- (void)viewDidUnload {
    
    _popController = nil;
    [super viewDidUnload];

}

- (void)popoverControllerDidDismissPopover:(UIPopoverController *)popoverController{

    popoverController = nil;
    NSLog(@"PopOver NIL");

}

-(UIImage*)editedImageFromMediaWithInfo:(NSDictionary*)info{
   
    UIImage *resultImage;
    
    //-(UIImage *)cropImage:(UIImage *)sourceImage cropRect:(CGRect)cropRect aspectFitBounds:(CGSize)finalImageSize fillColor:(UIColor *)fillColor {
    CGSize finalImageSize = CGSizeMake(700, 700);
    UIColor *fillColor = [UIColor clearColor];
    
    
    UIImage *sourceImage = [info objectForKey:UIImagePickerControllerOriginalImage];
    CGRect cropRect = [[info valueForKey:@"UIImagePickerControllerCropRect"] CGRectValue];
    CGImageRef sourceImageRef = sourceImage.CGImage;
    
    //Since the crop rect is in UIImageOrientationUp we need to transform it to match the source image.
    CGAffineTransform rectTransform = [self transformSize:sourceImage.size orientation:sourceImage.imageOrientation];
    CGRect transformedRect = CGRectApplyAffineTransform(cropRect, rectTransform);
    
    //Now we get just the region of the source image that we are interested in.
    CGImageRef cropRectImage = CGImageCreateWithImageInRect(sourceImageRef, transformedRect);
    
    //Figure out which dimension fits within our final size and calculate the aspect correct rect that will fit in our new bounds
    CGFloat horizontalRatio = finalImageSize.width / CGImageGetWidth(cropRectImage);
    CGFloat verticalRatio = finalImageSize.height / CGImageGetHeight(cropRectImage);
    CGFloat ratio = MIN(horizontalRatio, verticalRatio); //Aspect Fit
    CGSize aspectFitSize = CGSizeMake(CGImageGetWidth(cropRectImage) * ratio, CGImageGetHeight(cropRectImage) * ratio);
    
    
    CGContextRef context = CGBitmapContextCreate(NULL,
                                                 finalImageSize.width,
                                                 finalImageSize.height,
                                                 CGImageGetBitsPerComponent(cropRectImage),
                                                 0,
                                                 CGImageGetColorSpace(cropRectImage),
                                                 CGImageGetBitmapInfo(cropRectImage));
    
    if (context == NULL) {
        NSLog(@"NULL CONTEXT!");
    }
    
    //Fill with our background color
    CGContextSetFillColorWithColor(context, fillColor.CGColor);
    CGContextFillRect(context, CGRectMake(0, 0, finalImageSize.width, finalImageSize.height));
    
    //We need to rotate and transform the context based on the orientation of the source image.
    CGAffineTransform contextTransform = [self transformSize:finalImageSize orientation:sourceImage.imageOrientation];
    CGContextConcatCTM(context, contextTransform);
    
    //Give the context a hint that we want high quality during the scale
    CGContextSetInterpolationQuality(context, kCGInterpolationHigh);
    
    //Draw our image centered vertically and horizontally in our context.
    CGContextDrawImage(context, CGRectMake((finalImageSize.width-aspectFitSize.width)/2, (finalImageSize.height-aspectFitSize.height)/2, aspectFitSize.width, aspectFitSize.height), cropRectImage);
    
    //Start cleaning up..
    CGImageRelease(cropRectImage);
    
    CGImageRef finalImageRef = CGBitmapContextCreateImage(context);
    resultImage = [UIImage imageWithCGImage:finalImageRef];
    CGContextRelease(context);
    CGImageRelease(finalImageRef);
    
    return [UIImage imageWithData:UIImagePNGRepresentation(resultImage)];
    
}

- (CGAffineTransform) transformSize:(CGSize)imageSize orientation:(UIImageOrientation)orientation {
    
    CGAffineTransform transform = CGAffineTransformIdentity;
    switch (orientation) {
        case UIImageOrientationLeft: { // EXIF #8
            CGAffineTransform txTranslate = CGAffineTransformMakeTranslation(imageSize.height, 0.0);
            CGAffineTransform txCompound = CGAffineTransformRotate(txTranslate,M_PI_2);
            transform = txCompound;
            break;
        }
        case UIImageOrientationDown: { // EXIF #3
            CGAffineTransform txTranslate = CGAffineTransformMakeTranslation(imageSize.width, imageSize.height);
            CGAffineTransform txCompound = CGAffineTransformRotate(txTranslate,M_PI);
            transform = txCompound;
            break;
        }
        case UIImageOrientationRight: { // EXIF #6
            CGAffineTransform txTranslate = CGAffineTransformMakeTranslation(0.0, imageSize.width);
            CGAffineTransform txCompound = CGAffineTransformRotate(txTranslate,-M_PI_2);
            transform = txCompound;
            break;
        }
        case UIImageOrientationUp: // EXIF #1 - do nothing
        default: // EXIF 2,4,5,7 - ignore
            break;
    }
    return transform;
}


//EXTERNAL


// EXTERNAL
- (void)handleDocumentOpenURL:(NSString *)url {
    
    if (url) {
        PlayViewController *playViewController;
        
        NSData* data = [[NSData alloc] initWithContentsOfURL:[NSURL URLWithString:url]];
        UIImage *image = [UIImage imageWithData:data];
        data = nil;
        
        
        playViewController = [[PlayViewController alloc] initWithNibName:@"PlayViewController" bundle:nil];
        playViewController.userImage = image;
        
        [self presentViewController:playViewController animated:NO completion:nil];
    }
    
}


//UA
- (void)handleExternalURL:(NSString*)url{
    
    NSURL *URL = [NSURL URLWithString:url];
    
    [[UIApplication sharedApplication] openURL:URL];
    
    
}
- (void)handleInternalURL:(NSString*)url{
    
    NSURL *URL = [NSURL URLWithString:url];
    
    [[UIApplication sharedApplication] openURL:URL];
    
}

@end

File: CatWang2026/CatWang2026/stickerapp/ViewController.swift
//  I'ma Unicorn
//
//  Created by Franky Aguilar on 7/27/12.
//  Copyright (c) 2012 99centbrains Inc. All rights reserved.
//  @99centbrains - http://99centbrains.com
//  ALL ARTWORK AND DESIGN OWNED BY 99centbrains, not for reproduction or redistribution
//

import UIKit
import AVFoundation
import AudioToolbox

@objcMembers public class ViewController: UIViewController {
    
    // MARK: - Properties
    
    @IBOutlet weak var ibo_getphoto: UIView!
    
    private var popController: UIPopoverController?
    
    // MARK: - Lifecycle
    
    public override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    public override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        print("View Did Appear")
    }
    
    // MARK: - Photos
    
    @IBAction func iba_photoTake(_ sender: Any) {
        let picker = UIImagePickerController()
        picker.delegate = self
        picker.allowsEditing = true
        
        if UIImagePickerController.isSourceTypeAvailable(.camera) {
            picker.sourceType = .camera
            present(picker, animated: true, completion: nil)
        } else {
            iba_photoChoose(sender)
        }
    }
    
    @IBAction func iba_photoChoose(_ sender: Any) {
        let picker = UIImagePickerController()
        picker.delegate = self
        picker.allowsEditing = true
        picker.sourceType = .photoLibrary
        
        present(picker, animated: true, completion: nil)
    }
    
    @IBAction func iba_shareApp(_ sender: UIButton) {
        let shareText = kShareDescription
        guard let shareURL = URL(string: kShareURL) else { return }
        
        let activityItems: [Any] = [shareText, shareURL]
        let excludeActivities: [UIActivity.ActivityType] = [
            .addToReadingList,
            .assignToContact,
            .print
        ]
        
        let activityController = UIActivityViewController(
            activityItems: activityItems,
            applicationActivities: nil
        )
        activityController.excludedActivityTypes = excludeActivities
        
        activityController.completionWithItemsHandler = { activityType, completed, returnedItems, error in
            // Handle completion if needed
        }
        
        if UIDevice.current.userInterfaceIdiom == .pad {
            popController = UIPopoverController(contentViewController: activityController)
            popController?.contentSize = CGSize(width: 500, height: 425)
            let rect = sender.convert(sender.bounds, to: self.view)
            popController?.present(from: rect, in: self.view, permittedArrowDirections: .up, animated: true)
            SVProgressHUD.dismiss()
        } else {
            SVProgressHUD.dismiss()
            present(activityController, animated: true, completion: nil)
        }
    }
    
    @IBAction func iba_Facebook(_ sender: Any) {
        guard let url = URL(string: "http://facebook.com/99centbrains") else { return }
        let modalWeb = SVModalWebViewController(url: url)!
        present(modalWeb, animated: true, completion: nil)
    }
    
    @IBAction func iba_Twitter(_ sender: Any) {
        guard let url = URL(string: "http://twitter.com/catgnaw") else { return }
        let modalWeb = SVModalWebViewController(url: url)!
        present(modalWeb, animated: true, completion: nil)
    }
    
    @IBAction func iba_Insta(_ sender: Any) {
        guard let url = URL(string: "http://instagram.com/catgnaw") else { return }
        let modalWeb = SVModalWebViewController(url: url)!
        present(modalWeb, animated: true, completion: nil)
    }
    
    @IBAction func iba_Web(_ sender: Any) {
        guard let url = URL(string: "http://catgnaw.com") else { return }
        let modalWeb = SVModalWebViewController(url: url)!
        present(modalWeb, animated: true, completion: nil)
    }
    
    // MARK: - External URL Handling
    
    func handleDocumentOpenURL(_ url: String) {
        guard !url.isEmpty, let imageURL = URL(string: url) else { return }
        
        do {
            let data = try Data(contentsOf: imageURL)
            if let image = UIImage(data: data) {
                let storyboard = UIStoryboard(name: "MainStoryboard", bundle: nil)
                if let playViewController = storyboard.instantiateViewController(withIdentifier: "seg_PlayViewController") as? PlayViewController {
                    playViewController.userImage = image
                    present(playViewController, animated: false, completion: nil)
                }
            }
        } catch {
            print("Error loading image: \(error)")
        }
    }
    
    func handleExternalURL(_ url: String) {
        guard let urlToOpen = URL(string: url) else { return }
        UIApplication.shared.open(urlToOpen, options: [:], completionHandler: nil)
    }
    
    func handleInternalURL(_ url: String) {
        guard let urlToOpen = URL(string: url) else { return }
        UIApplication.shared.open(urlToOpen, options: [:], completionHandler: nil)
    }
    
    // MARK: - Image Editing
    
    private func editedImage(fromMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) -> UIImage? {
        let finalImageSize = CGSize(width: 700, height: 700)
        let fillColor = UIColor.clear
        
        guard let sourceImage = info[.originalImage] as? UIImage,
              let cropRectValue = info[.cropRect] as? CGRect,
              let sourceImageRef = sourceImage.cgImage else {
            return nil
        }
        
        let cropRect = cropRectValue
        
        // Transform crop rect to match source image orientation
        let rectTransform = transformSize(sourceImage.size, orientation: sourceImage.imageOrientation)
        let transformedRect = cropRect.applying(rectTransform)
        
        // Get the cropped region
        guard let cropRectImage = sourceImageRef.cropping(to: transformedRect) else {
            return nil
        }
        
        // Calculate aspect fit size
        let horizontalRatio = finalImageSize.width / CGFloat(cropRectImage.width)
        let verticalRatio = finalImageSize.height / CGFloat(cropRectImage.height)
        let ratio = min(horizontalRatio, verticalRatio)
        let aspectFitSize = CGSize(
            width: CGFloat(cropRectImage.width) * ratio,
            height: CGFloat(cropRectImage.height) * ratio
        )
        
        // Create bitmap context
        guard let colorSpace = cropRectImage.colorSpace,
              let context = CGContext(
                data: nil,
                width: Int(finalImageSize.width),
                height: Int(finalImageSize.height),
                bitsPerComponent: cropRectImage.bitsPerComponent,
                bytesPerRow: 0,
                space: colorSpace,
                bitmapInfo: cropRectImage.bitmapInfo.rawValue
              ) else {
            print("NULL CONTEXT!")
            return nil
        }
        
        // Fill with background color
        context.setFillColor(fillColor.cgColor)
        context.fill(CGRect(x: 0, y: 0, width: finalImageSize.width, height: finalImageSize.height))
        
        // Rotate and transform context based on source image orientation
        let contextTransform = transformSize(finalImageSize, orientation: sourceImage.imageOrientation)
        context.concatenate(contextTransform)
        
        // Set high quality interpolation
        context.interpolationQuality = .high
        
        // Draw image centered
        let drawRect = CGRect(
            x: (finalImageSize.width - aspectFitSize.width) / 2,
            y: (finalImageSize.height - aspectFitSize.height) / 2,
            width: aspectFitSize.width,
            height: aspectFitSize.height
        )
        context.draw(cropRectImage, in: drawRect)
        
        // Create final image
        guard let finalImageRef = context.makeImage() else {
            return nil
        }
        
        let resultImage = UIImage(cgImage: finalImageRef)
        
        // Convert to PNG data and back to ensure consistency
        guard let pngData = resultImage.pngData() else {
            return resultImage
        }
        
        return UIImage(data: pngData)
    }
    
    private func transformSize(_ imageSize: CGSize, orientation: UIImage.Orientation) -> CGAffineTransform {
        var transform = CGAffineTransform.identity
        
        switch orientation {
        case .left: // EXIF #8
            let txTranslate = CGAffineTransform(translationX: imageSize.height, y: 0.0)
            let txCompound = txTranslate.rotated(by: .pi / 2)
            transform = txCompound
            
        case .down: // EXIF #3
            let txTranslate = CGAffineTransform(translationX: imageSize.width, y: imageSize.height)
            let txCompound = txTranslate.rotated(by: .pi)
            transform = txCompound
            
        case .right: // EXIF #6
            let txTranslate = CGAffineTransform(translationX: 0.0, y: imageSize.width)
            let txCompound = txTranslate.rotated(by: -.pi / 2)
            transform = txCompound
            
        case .up, .upMirrored, .downMirrored, .leftMirrored, .rightMirrored:
            // EXIF #1 - do nothing, or ignore mirrored cases
            break
            
        @unknown default:
            break
        }
        
        return transform
    }
}

// MARK: - UIImagePickerControllerDelegate

extension ViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    
    public func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        if UIDevice.current.userInterfaceIdiom == .pad {
            popController?.dismiss(animated: true)
        }
        picker.dismiss(animated: true, completion: nil)
    }
    
    public func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
        let storyboard = UIStoryboard(name: "MainStoryboard", bundle: nil)
        
        guard let playViewController = storyboard.instantiateViewController(withIdentifier: "seg_PlayViewController") as? PlayViewController else {
            picker.dismiss(animated: true, completion: nil)
            return
        }
        
        picker.dismiss(animated: true) { [weak self] in
            playViewController.userImage = self?.editedImage(fromMediaWithInfo: info)
            self?.navigationController?.pushViewController(playViewController, animated: true)
        }
    }
}

// MARK: - UIPopoverControllerDelegate

extension ViewController: UIPopoverControllerDelegate {
    
    public func popoverControllerDidDismissPopover(_ popoverController: UIPopoverController) {
        popController = nil
        print("PopOver NIL")
    }
}
